testdata/p4tc_samples_outputs/simple_exact_example_control_blocks.c

```diff
@@ -76,8 +76,8 @@ static __always_inline int process(struct __sk_buff *skb, struct my_ingress_head
                     switch (value->action) {
                         case INGRESS_NH_TABLE_ACT_INGRESS_SEND_NH: 
                             {
-                                hdr->ethernet.srcAddr = value->u.ingress_send_nh.smac;
-                                                                hdr->ethernet.dstAddr = value->u.ingress_send_nh.dmac;
+                                hdr->ethernet.srcAddr = bpf_cpu_to_be64(value->u.ingress_send_nh.smac);
```

Comment:
Mac addresses are already in network order in the P4TC side, so there is no need to convert it here

Reply:
Is it only for macaddr or all fields received from lookup (i.e value->xxx) are in network order?

Reply:
It depends on the type of the action parameter
If it is `ipv6`, `ipv4`, `mac address`, or any of the `beXXX` types it's already in network order

Reply:
@vbnogueira  From language point of view , 
1. we assume a particular order of parameter/ argument (say network order or host order)
2. if this assumption is broken then we should use annotation to educate compiler say @tc_order("new_order") this new_order could be host or network.

Deducing order from type (mac/ ip4/ ip6 address) etc. is not scalable and compiler is not expected to have knowledge about these.

If a mac address is to be treated differently from other types(say all other are host and mac is network) 
then have a rule all objects >= xbyte will be network or say @tc_order is network.

using @tc_order is scalable where as other one is not. 
Pl suggest

Reply:
> @vbnogueira From language point of view ,
> 
> 1. we assume a particular order of parameter/ argument (say network order or host order)
> 2. if this assumption is broken then we should use annotation to educate compiler say @tc_order("new_order") this new_order could be host or network.

Sorry, I don't follow
We already have an annotation
In this case, specifically, it would be @tc_type("macaddr")
You're saying that semantically it being a type annotation complicates things?

> Deducing order from type (mac/ ip4/ ip6 address) etc. is not scalable and compiler is not expected to have knowledge about these.
> 
> If a mac address is to be treated differently from other types(say all other are host and mac is network) then have a rule all objects >= xbyte will be network or say @tc_order is network.

This won't work, at least for mac addresses, because u64, for example would have a byte size bigger than mac address and will be in host order

> using @tc_order is scalable where as other one is not. Pl suggest

So, we already are using an annotation for these particular types which is `@tc_type`
We could add this other one, but it would look a bit worse I think
If you take a look at simple_exact_example.p4 (https://github.com/Sosutha/p4c/blob/p4tc_default_action_fix/testdata/p4tc_samples/simple_exact_example.p4#L80), the action `send_nh's` definition would end up looking like this:

```
action send_nh(@tc_type("dev") PortId_t port_id, @tc_order @tc_type("macaddr") bit<48> dmac, @tc_order @tc_type("macaddr") bit<48> smac)
```

So `dmac` and `smac` now would have two annotations instead of one. That is, of course, if I'm understanding your suggestion correctly

Note that we also require `@tc_type("macaddr")` for the compiler to generate the type `macaddr` for the action parameters in the JSON

Reply:
@vbnogueira  I see type is being used for multiple purpose, 
1. type (non basic) like dev, ipv4, ipv6, mac which are used as type(in backend) and also infer endian(here size and endian both implicit)
2. basic type be16, be32 etc. where size and endian both are explicit.
 These 2 are mixed and can be confusing.
If you explicitly document non-basic type for their size./ usage and endian order it should be fine.

Just 1 other query can their be different endian for (ip4/ ipv6/ mac) then what is implicit currently(network order), and how that can be represented.

Reply:
Here's what i would consider as "documentation". We can improve it if needed.

First assumption from P4 if i am not mistaken is every header is network order. At least none of the P4 specs draws a distinction on  endianness. If i am not mistaken metadata is assumed to be "host order".

Endianness should be only to be very specific primitive types: 16/32/64/128 bit types. So no other types should be considered.
A MAC address being 48 bits is outside the scope of this, so requires no special interpretation. IOW, if you are doing computation on a MAC address you do not need to convert to host order.

The default assumption on headers should be for the 4 types that we always default to them being network order. Unless otherwise overridden. There are certain protocols, as i have mentioned in the past, that may have host order headers. If you have a very specific setup, example you know for sure your environment is x86, then you can have your protocol use host order headers. P4 should allow that and our generated code should be able to deal with that. Google's early implementation of QUIC infact did exactly that. Then your arithmetic computation can be done without any conversion. Annotations to indicate the header is to be overriden as host order is needed.
We have created types u16/be16, u32/be32, u64/be64, u128/be128 to describe these types - and annotate as such..
 
To the second point you mention @syogender :
For some of the types we create **_aliases_** to, e.g type "ipv4" is underneath per P4TC parlance be32. The alias "ipv4" helps us to provide utilities in the control plane such as how to parse these from the command line, how to print them etc. Same applies to the type alias "ipv6" (being a be128). The other types as "dev" or "macaddr" etc are also useful for the same reasons.

For metadata we can assume host order unless otherwise overriden. For example in our case the skb metadata called protocol is a be32 - we annotate as such.

Reply:
@komaljai per your explanation of the code and things entangled and given our sense our urgency we'd like to come up with a short term solution. We can then optimize in the future.  our email exchange, here's our current thinking (you may have a different view):

1. Restore Tomasz original approach where he converts everything to host order
2. Whenever a table key (that is a header field) has a tc_type annotation that implies network order(ipv4, ipv6, and beXX), convert the key to network order
3. Whenever a table key (that is a header field) has annotation tc_type("macaddr"), "convert" (it will be stored as a u64) to host order before doing the lookup
4. Whenever an action parameter has a tc_type annotation that implies network order(ipv4, ipv6, and beXX), convert the action parameter to host order after retrieving it from P4TC
5. Whenever an action parameter  has annotation tc_type("macaddr"), "convert" (it will be stored as a u64) to host order and shift it by 16 to the right after retrieving it from P4TC
6. In the case of add_entry that assigns an action, whenever an action parameter has a tc_type annotation that implies network order(ipv4, ipv6, and beXX), convert the action parameter to network order before putting it in action structure
7. In the case of add_entry that assigns an action, whenever an action parameter has annotation tc_type("macaddr"), convert the action parameter to network order (it will be stored as a u64) before copying it to the action parameter struct

I know you had other suggestions to make things cleaner but double check if this doable in the short term.
@vbnogueira if anything else please add..

Reply:
Addressed.